### PR TITLE
Replace paperclip with CarrierWave

### LINKED ISSUE #85: Replace paperclip with maintained alternative
[Paperclip](https://github.com/thoughtbot/paperclip) has been deprecated.

So, for file uploads and processing, we should switch to something that is actually maintained.

Two alternatives for this are:
- [Rails' `ActiveStorage`](https://guides.rubyonrails.org/active_storage_overview.html)
- [CarrierWave](https://github.com/carrierwaveuploader/carrierwave)

### PR DESCRIPTION
**Pull request by raccube**
Not to be merged until #68 has been merged. This PR omits theming as it is being reworked.

Closes #85

### CODE REVIEW

File: db/migrate/20141229085904_add_attachment_profile_picture_to_users.rb [hunk 1]
```diff
@@ -1,11 +1,17 @@
 class AddAttachmentProfilePictureToUsers < ActiveRecord::Migration[4.2]
   def self.up
     change_table :users do |t|
-      t.attachment :profile_picture
+      t.string :profile_picture_file_name
+      t.string :profile_picture_content_type
+      t.bigint :profile_picture_file_size
```
Comment (nilsding) [inline db/migrate/20141229085904_add_attachment_profile_picture_to_users.rb]: This was a `t.integer` before -- please change it (and create a migration for that if it really should become a `bigint`)

File: db/migrate/20150508144336_add_attachment_profile_header_to_users.rb [hunk 2]
```diff
@@ -1,7 +1,10 @@
 class AddAttachmentProfileHeaderToUsers < ActiveRecord::Migration[4.2]
   def change
     change_table :users do |t|
-      t.attachment :profile_header
+      t.string :profile_header_file_name
+      t.string :profile_header_content_type
+      t.bigint :profile_header_file_size
```
Comment (nilsding) [inline db/migrate/20150508144336_add_attachment_profile_header_to_users.rb]: This was a `t.integer` before -- please change it (and create a migration for that if it really should become a `bigint`)

File: db/migrate/20150825073030_create_themes.rb [hunk 3]
```diff
@@ -30,7 +30,10 @@ def change
       t.integer :background_text, limit: 4, default: 0x222222
       t.integer :background_muted, limit: 4, default: 0xBBBBBB
 
-      t.attachment :css
+      t.string :css_file_name
+      t.string :css_content_type
+      t.bigint :css_file_size
```
Comment (nilsding) [inline db/migrate/20150825073030_create_themes.rb:35]: This was a `t.integer` before -- please change it (and create a migration for that if it really should become a `bigint`)

File: config/initializers/25_carrierwave.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+CarrierWave.configure do |config|
+  config.fog_credentials = {
+    :provider => "Local",
+    :local_root => "#{Rails.root}/public",
+  }
+  config.fog_directory = "/system"
+
+  unless APP_CONFIG["fog"].nil?
+    config.fog_credentials = APP_CONFIG["fog"]["credentials"] unless APP_CONFIG["fog"]["credentials"].nil?
```
Comment (nilsding) [inline config/initializers/25_carrierwave.rb]: We are running Ruby 2.3+ now, so we can use `Hash#dig` here. 

```suggestion
    config.fog_credentials = APP_CONFIG.dig("fog", "credentials") unless APP_CONFIG.dig("fog", "credentials").nil?
```

File: config/environments/development.rb [hunk 5]
```diff
@@ -29,14 +29,8 @@
   # Action Mailer Setup
   # if you want to test sending mails locally, uncomment the line below and comment the :sendmail line
   # config.action_mailer.delivery_method = :letter_opener
-  if ENV["mailcatcher"] == "yes"
```
Comment (nilsding) [inline config/environments/development.rb]: Why was this changed?

File: app/uploaders/profile_picture_uploader.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  version :large do
+    process resize_to_fit: [500, 500]
+  end
+  version :medium do
+    process resize_to_fit: [256, 256]
+  end
+  version :small do
+    process resize_to_fit: [80, 80]
+  end
```
Comment (nilsding) [inline app/uploaders/profile_picture_uploader.rb:14]: Do these crop the pictures as well?  If not, can these be changed to allow for cropping?

(See also: [lib/paperclip_processors/cropper.rb](https://github.com/Retrospring/retrospring/blob/1d7430c3f786f42bb5e416eb4596f426d57a19a9/lib/paperclip_processors/cropper.rb), this is what the `processors: [:cropper]` option in the `has_attached_file` command ran)

File: app/uploaders/profile_header_uploader.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  version :web do
+    process resize_to_fit: [1500, 350]
+  end
+  version :mobile do
+    process resize_to_fit: [450, 105]
+  end
+  version :retina do
+    process resize_to_fit: [900, 210]
+  end
```
Comment (nilsding) [inline app/uploaders/profile_header_uploader.rb:14]: Likewise, do these crop the image?

File: Gemfile [hunk 8]
```diff
@@ -29,8 +29,6 @@ gem 'bootstrap_form'
 gem 'font-kit-rails'
 gem 'nprogress-rails'
 gem 'font-awesome-rails', '~> 4.7.0'
-gem "paperclip", "~> 5.2"
-gem 'delayed_paperclip'
```
Comment (nilsding) [inline Gemfile:32]: Can CarrierWave be configured so that it processes the images in a background job?

**Review by nilsding — CHANGES_REQUESTED** [review]

File: config/environments/development.rb (hunk 5 above)
Comment (raccube) [reply config/environments/development.rb]: Comitted by accident

File: Gemfile (hunk 8 above)
Comment (raccube) [reply Gemfile:32]: Tried [carrierwave_backgrounder](https://github.com/lardawge/carrierwave_backgrounder) but it didn't seem to want to work and hasn't been updated in quite a while.

There is [carrierwave_direct](https://github.com/dwilkie/carrierwave_direct) which works through providing direct uploads to S3.

**Comment by raccube** [comment]
**Left to do:**
- [x] Figure out a way of making image processing asyncronous
- [x] Remove unneeded database fields left over from paperclip

File: db/migrate/20200517190138_rename_crop_fields.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
+class RenameCropFields < ActiveRecord::Migration[5.2]
+  def up
```
Comment (nilsding) [inline db/migrate/20200517190138_rename_crop_fields.rb]: IIRC `rename_column` is a reversible operation; meaning that you can rename `#up` to `#change` and get rid of the `#down` method.

File: app/uploaders/base_uploader.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+  storage :fog
+
+  # Store original size
+  version :original
+
+  # Process cropping on upload
+  process :cropping
+
+  def store_dir
```
Comment (nilsding) [inline app/uploaders/base_uploader.rb:14]: Hm, this method is already defined by Carrierwave's Paperclip compatibility module ...  https://github.com/carrierwaveuploader/carrierwave/blob/d41ad71ad71a813dddf47e750e5a8b5f5c8d4e0d/lib/carrierwave/compatibility/paperclip.rb#L74

Any reason why this is required here?

File: db/migrate/20200517190138_rename_crop_fields.rb (hunk 9 above)
Comment (raccube) [reply db/migrate/20200517190138_rename_crop_fields.rb]: Neat!

File: app/uploaders/base_uploader.rb (hunk 10 above)
Comment (raccube) [reply app/uploaders/base_uploader.rb:14]: It's getting overriden to match the format with what we had before

**Review by nilsding — APPROVED** [review]
